### PR TITLE
combine scheduler cmd config into cmd options

### PR DESCRIPTION
**Pull request by mlmhl**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

This is a code clean PR: Move the `Config` definition from `cmd/kube-scheduler/app/config` into `cmd/kube-scheduler/app/options`.

**Which issue(s) this PR fixes**:

Fixes #71861 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by wgliang** [comment]
/lgtm
A good merge combination. Thanks @mlmhl

**Review by misterikkit — CHANGES_REQUESTED** [review]
Thanks for taking on this task. 

As part of the package merge, I am hoping to eliminate some redundant types. The `options` package now contains the following types,
1. `Options`
1. `Config`
1. `CompletedConfig`
1. `completedConfig`

Ideally, I would prefer to delete 2 through 4 and keep only 1. In particular, the only purpose of "completed" config is to apply some default values to fields. That can easily be done in `NewOptions`.

Because of the new scheduler constructor from phase 1, `cmd/kube-scheduler` does not need to use any "config" struct to create a scheduler instance.

File: cmd/kube-scheduler/app/server.go
Comment (misterikkit) [inline cmd/kube-scheduler/app/server.go]: I think this should be `cz.Set(opts.ComponentConfig)` to keep the current functionality.

File: cmd/kube-scheduler/app/server.go
Comment (misterikkit) [inline cmd/kube-scheduler/app/server.go]: It seems to me that the fields which were moved from `Config` into `Options` are written once in `initialize()` and read once right here. We could probably have them as local variables in this function, rather than keeping them in the `Options` struct. What do you think?

**Review by misterikkit — CHANGES_REQUESTED** [review]

File: cmd/kube-scheduler/app/options/options.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/options/options.go]: I read the code again and found that `deprecated` for this field is not correct, as we apply `DelegatingAuthenticationOptions` to `AuthenticationInfo` and then use `AuthenticationInfo` to create the `Healthz` handler. So I will remove the `Deprecated` prefix for this filed.

BTW, `AuthorizationInfo` and `SecureServingInfo` are similar.

File: cmd/kube-scheduler/app/options/options.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/options/options.go]: Done

File: cmd/kube-scheduler/app/server.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/server.go]: Done

File: cmd/kube-scheduler/app/options/options.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/options/options.go]: I found this problem too, so I explicit invoke `Options.Initialize` just before `app.Run`.

File: cmd/kube-scheduler/app/server.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/server.go]: I move the old `Config` fields to `app.Run` as local variables, except `SecureServingInfo`, `InsecureServingInfo` and `InsecureMetricsServingInfo`, the reasons are as follows:

1. [testserver](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-scheduler/app/testing/testserver.go) may use `LoopbackClientConfig` to create a k8s client before `app.Run`: https://github.com/kubernetes/kubernetes/blob/a78b591ed9fc42cb3f896d6a7f8178008dbf1239/cmd/kube-scheduler/app/testing/testserver.go#L129

that means we need to initialize `LoopbackClientConfig` inside `Options.Initialize`. Unfortunately, `InsecureServingInfo` and `InsecureMetricsServingInfo` are initialized along with `LoopbackClientConfig`: https://github.com/kubernetes/kubernetes/blob/a78b591ed9fc42cb3f896d6a7f8178008dbf1239/cmd/kube-scheduler/app/options/insecure_serving.go#L63

So change `InsecureServingInfo` and `InsecureMetricsServingInfo` as local variables makes initialization difficult to perform.

2. The integration test wants to verify `SecureServingInfo` and `InsecureServingInfo`, this requires these fields be exported: https://github.com/kubernetes/kubernetes/blob/a78b591ed9fc42cb3f896d6a7f8178008dbf1239/test/integration/serving/serving_test.go#L76

File: cmd/kube-scheduler/app/options/options.go
Comment (misterikkit) [reply cmd/kube-scheduler/app/options/options.go]: > BTW, `AuthorizationInfo` and `SecureServingInfo` are similar.

Do you think that we need both of them? I won't pretend to understand all the moving parts of apimachinery server libraries. The contents of those two structs looks fairly different to me. https://github.com/kubernetes/apiserver/blob/master/pkg/server/config.go#L202

File: cmd/kube-scheduler/app/server.go
Comment (misterikkit) [reply cmd/kube-scheduler/app/server.go]: That seems reasonable to me.

PS: How did you get the code snippets to show up in your comment?

File: cmd/kube-scheduler/app/options/options.go
Comment (misterikkit) [inline cmd/kube-scheduler/app/options/options.go]: What do you think about putting this whole func inline in `Initialize`? It seems a bit weird to have this function called from exactly one place.

File: cmd/kube-scheduler/app/options/options.go
Comment (misterikkit) [inline cmd/kube-scheduler/app/options/options.go]: Maybe we could add a comment that says this must be called after flags have been parsed?

File: cmd/kube-scheduler/app/server.go
Comment (misterikkit) [inline cmd/kube-scheduler/app/server.go]: > uniquifier

I like this word

File: cmd/kube-scheduler/app/testing/testserver.go
Comment (misterikkit) [inline cmd/kube-scheduler/app/testing/testserver.go]: optional: combine this line into the if-statement.

**Review by misterikkit — COMMENTED** [review]
I think you can remove WIP from this PR.

FYI, this PR (https://github.com/kubernetes/kubernetes/pull/71875) also relates to scheduler initialization.
I don't think there is a conflict, but there may be more work after both of these PRs to eliminate more "config" stuff.

File: cmd/kube-scheduler/app/options/options.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/options/options.go]: Done

File: cmd/kube-scheduler/app/testing/testserver.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/testing/testserver.go]: Done

File: cmd/kube-scheduler/app/options/options.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/options/options.go]: Done

File: cmd/kube-scheduler/app/server.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/server.go]: I just paste the code link in the comment. It seems that GitHub only show the code snippets for links of current repo(kubernetes). Code links of other repos(such as apiserver, apimachinery, etc.) won't be showed as snippets.

File: cmd/kube-scheduler/app/options/options.go
Comment (mlmhl) [reply cmd/kube-scheduler/app/options/options.go]: > Do you think that we need both of them?

I'm not sure if I understand what you mean. Did you mean that why we need both of `AuthenticationInfo`(which type is `apiserver.AuthorizationInfo`) and `Authentication`(which type is `apiserveroptions.DelegatingAuthenticationOptions`)? Or anything else?

File: cmd/kube-scheduler/app/server.go
Comment (Huang-Wei) [reply cmd/kube-scheduler/app/server.go]: @misterikkit try use "permanent link".

image

File: cmd/kube-scheduler/app/options/options_test.go
Comment (misterikkit) [inline cmd/kube-scheduler/app/options/options_test.go]: :+1:

**Review by misterikkit — APPROVED** [review]
/lgtm
/assign @bsalamat

**Review by bsalamat — COMMENTED** [review]
/approve

**Comment by misterikkit** [comment]
Looks like you need to resolve the conflict and then get approval from @kubernetes/sig-testing-pr-reviews

**Comment by mlmhl** [comment]
@misterikkit PR is rebased.

**Comment by misterikkit** [comment]
/assign @fejta 
Do you mind approving test changes?

**Comment by fejta** [comment]
Looks like we're missing a test/integration/serving/OWNERS file

/unassign
/assign @sttts 

Alternatively, please copy the test/integration/scheduling/OWNERS file to test/integration/serving and assign back to me to approve

**Comment by sttts** [comment]
/hold

The change goes against the patterns we are trying to establish in all of our components:
- **options**: basically the struct holding the variables bound against flags, plus maybe a few other fields used for custom plumbing.
- **config**: static (nothing is running yet) runtime objects derived from options, e.g. clients, informers, loaded certs
- **server**: the "running" data structure.

In some cases config and server are the same struct.

When I rememeber right, the scheduler had another layer between config and server. That should be cleaned up.

**Comment by misterikkit** [comment]
@sttts Thanks for the feedback. I have been lurking in the '#wg-component-standard' slack but I'm not familiar with the grand vision. I agree that keeping "options" separate and related only to flags is desirable and consistent with the rest of k8s. I had not seen this as a violation of that until you pointed it out.

I don't take issue with the idea of "config" in this context, but the way it gets used in practice bothers me. e.g. our "server" looks like this.

```go
type Scheduler struct {
    config factory.Config
}
```

You are right that the scheduler has layers which need to be removed. [Here](https://gist.githubusercontent.com/misterikkit/b5c5d56f55a52df48839f3e1aa5ec48b/raw/819aaf76a04a57a775c87b477e446d614b579965/K8S%2520Scheduler%2520Configuration%2520Handling.png) is my initial analysis of the current situation. ([context](https://groups.google.com//forum/#!topic/kubernetes-sig-scheduling/HVmEEG1zjyk))

But I also think that runtime state gets leaked into "config". Informers are runtime and not static, in my opinion. Somebody has to call `Start(stopCh)` and `close(stopCh)` on them.


Anyway, I have veered off topic. My framing device for this PR was originally, "what is the correct API between `cmd/kube-scheduler` and `pkg/scheduler`?" This is an area where hopefully WG-component-standard can give guidance. In https://github.com/kubernetes/kubernetes/issues/68954, I advocated the best golang practices I knew of. The resulting `scheduler.New` function does not consume any "config", but instead has required arguments followed by functional options.

Internally to `pkg/scheduler` I think we should get rid of any "config" objects since we set up a "server" and run it. Externally to `pkg/scheduler` I don't see the point of a "config" given the way we wrote `scheduler.New`.

**Comment by sttts** [comment]
@misterikkit we are usually not very strict with splitting config and server, i.e. config will have clients, will have informers. My rule of thumb usually is that config should not have something that is started or run (with a `stopCh`) already, i.e. a config should be "unstarted". Also config can be copied and reused, started server state can't. When the objects are started, the config turns into the server state. 

Trying to establish that rule in a big plumbing code base usually leads a structure which is much easier to understand than without the rule, although it might be hard at times.

The functional options pattern certainly can replace configs. I could imagine that the `options.ApplyTo` just produces those functional options. Let's call them functional config to avoid confusion. I.e.:

```
type ConfigFunc func(*Server)

func (o *Options) ToConfig() ([]ConfigFunc, error)
```

What do you think? Could be a nice pattern, maybe a better alternative to our giant configs.

**Comment by vllry** [comment]
LGTM for what it's worth.

@mlmhl are you able to handle the merge conflicts to get this merged?

**Comment by ingvagabund** [comment]
Given https://github.com/kubernetes/kubernetes/issues/71861 was closed, there has not been any response from @mlmhl for more than a year (https://github.com/kubernetes/kubernetes/issues/71861#issuecomment-522631333) and the underlying code has be non-trivially modified, I think this issue can be closed.

/close

@mlmhl feel free to re-open if you think otherwise

**Pull request closed without merging**
